File: agent-service/src/agents/coordinator/drivers/githubCopilot.ts
import type { CodingAgentDriver, AgentTask, AgentResult, AgentStatus } from "./base";
import type { Skill } from "../../../core/types";

export interface GitHubCopilotDriverConfig {
  githubToken?: string;
  llmProvider?: string;
  llmApiKey?: string;
  llmBaseUrl?: string;
  skills: Skill[];
}

export class GitHubCopilotDriver implements CodingAgentDriver {
  private status: AgentStatus = "idle";
  private config: GitHubCopilotDriverConfig;

  constructor(cfg: GitHubCopilotDriverConfig) {
    this.config = cfg;
  }

  async getStatus(): Promise<AgentStatus> {
    return this.status;
  }

  async execute(task: AgentTask, workspaces: Record<string, string>): Promise<AgentResult> {
    this.status = "busy";
    const primaryWorkspace = Object.values(workspaces)[0] ?? process.cwd();

    try {
      const authOptions = this.config.githubToken
        ? { token: this.config.githubToken }
        : {
            byok: {
              provider: this.config.llmProvider ?? "openai",
              apiKey: this.config.llmApiKey ?? "",
              ...(this.config.llmBaseUrl ? { baseURL: this.config.llmBaseUrl } : {}),
            },
          };

      // @github/copilot-sdk — installed separately: npm install @github/copilot-sdk
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const copilotModule = require("@github/copilot-sdk") as {
        createClient: (auth: unknown) => { run: (opts: { prompt: string; cwd: string; tools: string[] }) => Promise<{ output?: string }> }
      };
      const client = copilotModule.createClient(authOptions);

      const skillsContext =
        this.config.skills.length > 0
          ? "\n\nAvailable skills:\n" +
            this.config.skills
              .map((s) => `## ${s.name}\n${s.description}\n${s.instructions}`)
              .join("\n\n")
          : "";

      const prompt = `${task.title}\n\n${task.description}${skillsContext}`;

      const result = await client.run({
        prompt,
        cwd: primaryWorkspace,
        tools: ["all"],
      });

      this.status = "idle";
      return { success: true, output: result?.output ?? "Done" };
    } catch (err) {
      this.status = "error";
      return { success: false, output: "", error: String(err) };
    }
  }
}

File: agent-service/src/agent_service/agents/coordinator/drivers/github_copilot.py
from __future__ import annotations

from typing import List, Optional

from agent_service.agents.coordinator.drivers.base import AgentArtifact, AgentResult, AgentStatus, AgentTask
from agent_service.core.types import Skill


class GitHubCopilotDriver:
    """Driver that uses the GitHub Copilot SDK for task execution."""

    def __init__(
        self,
        github_token: Optional[str],
        llm_provider: Optional[str],
        llm_api_key: Optional[str],
        llm_base_url: Optional[str],
        skills: List[Skill],
    ) -> None:
        self._github_token = github_token
        self._llm_provider = llm_provider
        self._llm_api_key = llm_api_key
        self._llm_base_url = llm_base_url
        self._skills = skills
        self._status: AgentStatus = "idle"

    async def get_status(self) -> AgentStatus:
        return self._status

    async def execute(self, task: AgentTask, workspaces: dict[str, str]) -> AgentResult:
        self._status = "busy"
        primary_workspace = next(iter(workspaces.values()), ".")

        try:
            import importlib

            copilot = importlib.import_module("github_copilot_sdk")

            auth_options = (
                {"token": self._github_token}
                if self._github_token
                else {
                    "byok": {
                        "provider": self._llm_provider or "openai",
                        "apiKey": self._llm_api_key or "",
                        **({"baseURL": self._llm_base_url} if self._llm_base_url else {}),
                    }
                }
            )

            client = copilot.create_client(auth_options)

            skills_context = ""
            if self._skills:
                skills_context = "\n\nAvailable skills:\n" + "\n\n".join(
                    f"## {s.name}\n{s.description}\n{s.instructions}" for s in self._skills
                )

            result = await client.run(
                prompt=f"{task.title}\n\n{task.description}{skills_context}",
                cwd=primary_workspace,
                tools=["all"],
            )

            self._status = "idle"
            return AgentResult(success=True, output=getattr(result, "output", None) or "Done")
        except Exception as exc:
            self._status = "error"
            return AgentResult(success=False, output="", error=str(exc))
